pc-relate/pcrelate.wdl

```diff
@@ -0,0 +1,395 @@
+			f.write('variant_include_file "~{variant_include_file}"\n')
+		if "~{out_prefix}" != "":
+			f.write('out_prefix "~{out_prefix}"\n')
+		# n_pcs has a default (not SB default!) in CWL ergo always is defined so we don't check here
+		f.write("n_pcs ~{n_pcs}\n")
+		if "~{defSampleInclude}" == "true":
+			f.write('sample_include_file "~{sample_include_file}"\n')
+		f.write("n_sample_blocks ~{n_sample_blocks}\n")
+		f.write("ibd_probs ~{ibd_probs}\n")
+		f.close()
+		CODE
+
+		R -q --vanilla --args pcrelate.config --segment ~{segment} < /usr/local/analysis_pipeline/R/pcrelate.R
+
+	}
+	
+	runtime {
+		cpu: cpu
+		docker: "uwgac/topmed-master@sha256:0bb7f98d6b9182d4e4a6b82c98c04a244d766707875ddfd8a48005a9f5c5481e"
+		disks: "local-disk " + finalDiskSize + " HDD"
+		memory: "${memory} GB"
+		preemptibles: "${preempt}"
+	}
+	output {
+		# CWL seems to have inconsistency where this is marked an optional output
+		# but it also is a not-optional input elsewhere -- double check!
+		File block = glob("*.RData")[0]
+	}
+}
+
+task pcrelate_correct {
+	input {
+		Array[File] pcrelate_block_files
+		Int n_sample_blocks = 1
+		Float sparse_threshold = 0.02209709
+
+		# runtime attributes
+		Int addldisk = 1
+		Int cpu = 2
+		Int memory = 4
+		Int preempt = 3
+	}
+	# Estimate disk size required
+	Int block_size = ceil(size(pcrelate_block_files, "GB"))
+	Int finalDiskSize = 2*block_size + addldisk
+
+	command <<<
+		set -eux -o pipefail
+
+		BASH_FILES=(~{sep=" " pcrelate_block_files})
+		for BASH_FILE in ${BASH_FILES[@]};
+		do
+			ln -s ${BASH_FILE} .
+		done
+
+		python << CODE
+		import os
+		f = open("pcrelate_correct.config", "a")
+		one_file = ["~{sep="," pcrelate_block_files}"][0]
+		prefix = os.path.basename(one_file).split("_block_")[0]
+		f.write("pcrelate_prefix \"%s\"\n" % prefix)
+		f.write("n_sample_blocks ~{n_sample_blocks}\n")
+		f.write("sparse_threshold ~{sparse_threshold}\n")
+		CODE
+
+		R -q --vanilla --args pcrelate_correct.config < /usr/local/analysis_pipeline/R/pcrelate_correct.R
+
+	>>>
+	
+	runtime {
+		cpu: cpu
+		docker: "uwgac/topmed-master@sha256:0bb7f98d6b9182d4e4a6b82c98c04a244d766707875ddfd8a48005a9f5c5481e"
+		disks: "local-disk " + finalDiskSize + " HDD"
+		memory: "${memory} GB"
+		preemptibles: "${preempt}"
+	}
+	output {
+		# CWL seems to have inconsistency where this is marked an optional output
+		# but it also is a not-optional input into kinship_plots -- double check!
+		File pcrelate_output = glob("*_pcrelate.RData")[0]
+		File pcrelate_matrix = glob("*_pcrelate_Matrix.RData")[0]
+	}
+}
+
+task kinship_plots {
+	input {
+		File kinship_file
+		String? kinship_method = "pcrelate"  # hardcoded on CWL?
+		Float? kinship_plot_threshold
+		File? phenotype_file
+		String? group
+		File? sample_include_file
+		String out_prefix_initial = ""
+		
+		# runtime attributes
+		Int addldisk = 1
+		Int cpu = 2
+		Int memory = 4
+		Int preempt = 3
+	}
+	String out_prefix_final = out_prefix_initial + "_pcrelate"
```

Comment:
It is not clear why we added this to the WDL. I don't see it in the CWL.

Reply:
I see now that it is in the workflow level. Can you move this call there?

Reply:
I can, but doing so would require passing it in as an input into the task, and that would allow users to overwrite it and potentially break it.

Reply:
To clarify...

The CWL takes out_prefix, which the user can set or just leave blank, then slaps on "_pcrelate". The user cannot remove that second part. It's there whether they set out_prefix or not. This does indeed mean that not setting that value will result in the output of this task, and this task only (as the others don't mess with out_prefix), will be stuff like "_pcrelate_cross_group.pdf"

In WDL, we cannot pass variables into tasks unless they are public input variables (afaik), ie, are in the input section. But anything in the input section can be overwritten by the user. For instance, even though addldisk is not a workflow level input, it's a task level input that a user can overwrite and it gets its own box on Terra, because it's in the input section and everything in that section is public.

So, if out_prefix_final were in the workflow section, it would need to be passed in as an input and therefore be exposed. The user could then overwrite it. As a result, a user could end up with "dinosaurs_cross_group.pdf" on the WDL which is impossible to get on the CWL.